source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,296 @@
+    case SpvOpAtomicCompareExchangeWeak:
+      if (operand_index != 0 && operand_index != 1) {
+        return false;
+      }
+      break;
+
+    default:
+      return false;
+  }
+  return true;
+}
+
+bool TransformationChangingMemorySemantics::IsValidConverstion(
+    SpvOp opcode, SpvMemorySemanticsMask first_5bits_old_memory_semantics,
+    SpvMemorySemanticsMask first_5bits_new_memory_semantics,
+    SpvMemoryModel memory_model) {
+  std::set<SpvMemorySemanticsMask> atomic_load_set{
+      SpvMemorySemanticsMaskNone, SpvMemorySemanticsAcquireMask,
+      SpvMemorySemanticsSequentiallyConsistentMask};
+
+  std::set<SpvMemorySemanticsMask> atomic_store_set{
+      SpvMemorySemanticsMaskNone, SpvMemorySemanticsReleaseMask,
+      SpvMemorySemanticsSequentiallyConsistentMask};
+
+  std::set<SpvMemorySemanticsMask> atomic_rmw_set{
+      SpvMemorySemanticsMaskNone, SpvMemorySemanticsAcquireReleaseMask,
+      SpvMemorySemanticsSequentiallyConsistentMask};
+
+  std::set<SpvMemorySemanticsMask> barrier_set{
+      SpvMemorySemanticsMaskNone, SpvMemorySemanticsAcquireMask,
+      SpvMemorySemanticsReleaseMask, SpvMemorySemanticsAcquireReleaseMask,
+      SpvMemorySemanticsSequentiallyConsistentMask};
+
+  if (first_5bits_new_memory_semantics ==
+          SpvMemorySemanticsSequentiallyConsistentMask &&
+      memory_model == SpvMemoryModelVulkan) {
+    return false;
+  }
+  if (first_5bits_old_memory_semantics >= first_5bits_new_memory_semantics) {
+    return false;
+  }
+  switch (opcode) {
+    case SpvOpAtomicLoad:
+      return (atomic_load_set.find(first_5bits_old_memory_semantics) !=
+                  atomic_load_set.end() &&
+              atomic_load_set.find(first_5bits_new_memory_semantics) !=
+                  atomic_load_set.end());
+
+    case SpvOpAtomicStore:
+      return (atomic_store_set.find(first_5bits_old_memory_semantics) !=
+                  atomic_store_set.end() &&
+              atomic_store_set.find(first_5bits_new_memory_semantics) !=
+                  atomic_store_set.end());
+
+    case SpvOpAtomicExchange:
+    case SpvOpAtomicIIncrement:
+    case SpvOpAtomicIDecrement:
+    case SpvOpAtomicIAdd:
+    case SpvOpAtomicISub:
+    case SpvOpAtomicSMin:
+    case SpvOpAtomicUMin:
+    case SpvOpAtomicSMax:
+    case SpvOpAtomicUMax:
+    case SpvOpAtomicAnd:
+    case SpvOpAtomicOr:
+    case SpvOpAtomicXor:
+    case SpvOpAtomicCompareExchange:
+    case SpvOpAtomicCompareExchangeWeak:
+
+      return (atomic_rmw_set.find(first_5bits_old_memory_semantics) !=
+                  atomic_rmw_set.end() &&
+              atomic_rmw_set.find(first_5bits_new_memory_semantics) !=
+                  atomic_rmw_set.end());
+
+    case SpvOpControlBarrier:
+    case SpvOpMemoryBarrier:
+    case SpvOpMemoryNamedBarrier:
+
+      if (first_5bits_old_memory_semantics == SpvMemorySemanticsAcquireMask &&
+          first_5bits_new_memory_semantics == SpvMemorySemanticsReleaseMask) {
+        return false;
+      }
```

Comment:
Add a comment to explain why we need this special case check. In fact, you should also add a comment at the top of the function body explaining what it is we are trying to check. In other words, the things we list in our doc:

* Load: None -> Acquire  -> SequentiallyConsistent
* Store: None -> Release -> SequentiallyConsistent
* RMW: None -> AcquireRelease -> SequentiallyConsistent
* Barrier: None -> Acquire / Release -> AcquireRelease -> SequentiallyConsistent
* When the memory model is Vulkan, SequentiallyConsistent must not be used.

Reply:
Done.

Reply:
I don't see a comment here.

Reply:
Done.